### PR TITLE
bootkube: Inject bootstrap MachineConfigs into cluster

### PR DESCRIPTION
**Pull request by cgwalters**
This is an ugly fix for
https://github.com/openshift/machine-config-operator/issues/367

### CODE REVIEW

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template [hunk 1]
```diff
@@ -201,6 +201,15 @@ echo "etcd cluster up. Killing etcd certificate signer..."
 
 podman rm --force etcd-signer
 rm --force /etc/kubernetes/manifests/machineconfigoperator-bootstrap-pod.yaml
+# Copy the bootstrap MCs to inject into the target cluster
+# Yes this is a brutal hack, need to improve the MCC bootstrap above
+# 9a so we're after 99 - should change the others to 50- or something?
+for x in /etc/mcs/bootstrap/machine-configs/*.yaml; do
```
Comment (wking) [inline data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: Where do these come from in the first place?  Looks like `/etc/mcs` has only been mentioned in 6da1973fb (#879), but I don't see any installer code obviously populating it.  Maybe it's [here][1]?  If so, is there a reason you couldn't address this entirely within openshift/machine-config-operator?

[1]: https://github.com/openshift/machine-config-operator/pull/226/files#diff-51bc64f09fd6698cbb13c264c6b29e24R13

Comment (cgwalters) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: My understanding of the current code is here https://github.com/openshift/machine-config-operator/issues/367#issuecomment-460368011

> but I don't see any installer code obviously populating it. 

AIUI it's the static pod which we are just removing above this code:
`/etc/kubernetes/manifests/machineconfigoperator-bootstrap-pod.yaml`

> If so, is there a reason you couldn't address this entirely within openshift/machine-config-operator?

Maybe - I didn't write this code and am still learning things here, so alternative suggestions appreciated!

Comment (cgwalters) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: As far as I understand things though...that bootstrap pod is what's serving out the Ignition to the masters.  We don't have masters online until after `echo "etcd cluster up. Killing etcd certificate signer..."` (right?)

I believe we won't even have the `machineconfigs` CRD in the cluster until the MCO comes online.  So we will end up waiting in `openshift.service` for that - similar to how we wait for the cluster API objects and inject those.

**Comment by abhinavdahiya** [comment]
> This is an ugly fix for
> [openshift/machine-config-operator#367](https://github.com/openshift/machine-config-operator/issues/367)

Why do you believe the bootstrap mcc and long-running mcc are creating 2 different final machineconfigs?

/hold

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template (hunk 1 above)
Comment (wking) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: > > but I don't see any installer code obviously populating it.
>
> AIUI it's the static pod which we are just removing above this code:
`/etc/kubernetes/manifests/machineconfigoperator-bootstrap-pod.yaml`

So can we update that code to make these alterations?  It's maybe [here][1] calling [here][2]?

> So we will end up waiting in `openshift.service` for that - similar to how we wait for the cluster API objects and inject those.

I'm not sure how this comes into this pull request, but I'm pushing to get `openshift.service` and `openshift.sh` functionality moved into [`cluster-bootstrap`][3].  See #1147, which I'm going to reroll after `cluster-bootstrap` picks up openshift/library-go#220.

[1]: https://github.com/openshift/machine-config-operator/blob/c5f479e7ae15dfb579c2cf2b1575ceba9d0098e0/pkg/controller/bootstrap/bootstrap.go#L100-L133
[2]: https://github.com/openshift/machine-config-operator/blob/c5f479e7ae15dfb579c2cf2b1575ceba9d0098e0/pkg/controller/render/render_controller.go#L499-L548
[3]: https://github.com/openshift/cluster-bootstrap

**Comment by cgwalters** [comment]
> Clearly a lot of people are hitting this.

Just to emphasize this, look at e.g. [this recent PR to the installer](https://github.com/openshift/installer/pull/1181) which passed CI just fine...yet if you drill down into [the clusteroperator status](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1181/pull-ci-openshift-installer-master-e2e-aws/3383/artifacts/e2e-aws/clusteroperators.json) you'll notice the MCO is degraded:

`                    "master": "pool is degraded because of 1 nodes are reporting degraded status on update. Cannot proceed.",`

**Comment by cgwalters** [comment]
We need to get to the point where we're gating `e2e-aws` on nodes not being degraded (which is an interesting topic...I'd been adding tests for this to the MCO repo but do we have a way to extend e2e-aws with tests that live in other git repos besides origin)?

I fully admit this PR is a hack but I'm trying to make some progress on this as it's blocking my work on OS updates.

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template (hunk 1 above)
Comment (cgwalters) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: > So can we update that code to make these alterations?

Mmm...this code needs to generate content to inject into the target cluster.  The static pod is running on the bootstrap before the cluster, right?  

> but I'm pushing to get openshift.service and openshift.sh functionality moved into cluster-bootstrap.

That makes sense to me...I was surprised at the low-tech nature of `openshift.sh`.

**Comment by cgwalters** [comment]
While I think this could still make sense...we don't need this right now since we did end up identifying the underlying problem in https://github.com/openshift/installer/pull/1194

**Comment by cgwalters** [comment]
Reopening this; just hit the failure case again here: https://github.com/openshift/machine-config-operator/pull/363#issuecomment-462973819

I know this isn't the most beautiful code but it should let us avoid this problem; we will be able to see the config drift.

**Comment by abhinavdahiya** [comment]
> I know this isn't the most beautiful code but it should let us avoid this problem;

Open to any PRs that use mc* render to handle this like other operators. no inline.

**Comment by deads2k** [comment]
We have hit this in the field caused by a cluster-admin not getting all the options quite right. As I understand it, this can alleviate the "your cluster isn't installing" problem automatically in the unfortunate case and carries no negative side-effects in the common case.  We should get this into a state where it (or something like it) can merge. 

/reopen

**Comment by cgwalters** [comment]
Amazingly, this PR rebases cleanly.  However, looking at it now with months more of experience, I think we can clean this up...

**Comment by cgwalters** [comment]
Moved this to https://github.com/openshift/installer/pull/2936

**Comment by runcom** [comment]
> As I understand it, this can alleviate the "your cluster isn't installing" problem automatically in the unfortunate case

I don't fully agree with that, in the event there's a drift we're installing something which is what the user wants but then the MCO will reconcile to something completely not intended

**Comment by deads2k** [comment]
> > As I understand it, this can alleviate the "your cluster isn't installing" problem automatically in the unfortunate case
> 
> I don't fully agree with that, in the event there's a drift we're installing something which is what the user wants but then the MCO will reconcile to something completely not intended but the cluster can still function and nobody notices, it should be fatal instaed

reply in https://github.com/openshift/installer/pull/2936#issuecomment-575188306

**Pull request closed without merging**